FreeRTOS/FreeRTOS-Plus-TCP
# Avoid DHCP race

Description
-----------
[Andy Bennett](https://github.com/andyjpb) reported in [PR #123](https://github.com/FreeRTOS/FreeRTOS/pull/123) that sometimes the DHCP status machine keeps passing through same sequence of states again and again.

He wrote that it helps when `dhcpINITIAL_TIMER_PERIOD` is given a higher value.

Here is what happened:

While waiting for a DHCP reply, two events happen almost at the same time:

1) A UDP packet comes in with an acknowledgement
2) An internal DHCP timer expires ( after `dhcpINITIAL_TIMER_PERIOD` ticks ).

After event-1 the state goes to `eLeasedAddress`, which lasts until a time-out is reached.
After event-2, cause by a time-out, leasing is finished, and a new request will be sent.

Event-2 should be ignored because a new time-out has already been scheduled. Event-2 was still in the pipe-line, the working queue of the IP-task.

Solution: when sending an event to the DHCP driver, the current state will be sent as well. When receiving the event, the DHCP driver will check if the event is still applicable:

~~~c
    if( EP_DHCPData.eDHCPState != eExpectedState )
    {
        /* Ignore the event, it was queued and it comes too late. */
    }
    else
    {
        /* Let the DHCP state machine do one cycle. */
    }
~~~

Test Steps
-----------
Give the macro `dhcpINITIAL_TIMER_PERIOD` a very low value, about the time it takes to get a DHCP acknowledgement.
If it lasts e.g. 30 msec, you give it this value in FreeRTOSIPConfig.h.
~~~c
#define dhcpINITIAL_TIMER_PERIOD   pdMS_TO_TICKS( 30U )
~~~
It is difficult to get the race condition occur, I have seen it happening only a few times.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
